text/0000-formalise-reborrows.md

```diff
@@ -0,0 +1,197 @@
+that we get:
+
+```
+error[E0382]: use of moved value: `x`
+  --> src/main.rs:17:33
+   |
+16 |     a(x);
+   |       - value moved here
+17 |     println!("b called: a({})", x);
+   |                                 ^ value used here after move
+   |
+   = note: move occurs because `x` has type `MyRef<'_, str>`, which does not implement the `Copy` trait
+```
+
+The compiler is telling us that `a(x)` is interpreted as *copy `x` into `a` if
+`x` supports `Copy`, otherwise move `x` into `a`*. Except, as we know, if `x`
+had type `&str` instead of `MyRef<str>` for both the argument and parameter
+type, the compiler would have no problem interpreting this as a *reborrow*.
+A reborrow is not quite a copy due to the requirement for shorter lifetimes.
+
+Full example: https://play.rust-lang.org/?gist=d343ac07c9faf21607a5ad92bbaf5f45&version=stable
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Reborrowing trait
+
+We introduce the following trait, as member of `std::marker`:
+
+```rust
+trait Reborrow {
+    fn reborrow(&self) -> ???;
+}
+```
+
+Unfortunately the return type of `fn reborrow` isn't representable in general;
+the return type must be `Self` but with different lifetime(s) (and note that
+there will be multiple lifetimes if the type has multiple fields with
+lifetimes). It is not possible to use an associated type `type Result: ???`
+because the return type includes a lifetime (or multiple) bound by the caller.
+
+Fortunately it *is* possible to write specific implementations in Rust today,
+for example the above `MyRef` type can be [given a reborrow implementation](https://play.rust-lang.org/?gist=67e4600b48d123fea45dd2b9bdcbf035&version=stable):
+
+```rust
+impl<'a, T: 'a + ?Sized> Reborrow for MyRef<'a, T> {
+    fn reborrow<'b>(&self) -> MyRef<'b, T> where 'a: 'b {
+        MyRef(self.0)
+    }
+}
+```
+
+### Implementation
+
+The `&T` and `&mut T` types already support reborrowing semantics; this
+formalisation requires that both support the `Reborrow` trait.
+
+In theory, it should be possible to manually implement `Reborrow` with code
+similar to the above; I do not know if this is possible due to the improper
+definition of the return type.
+
+Rust should have built-in support for deriving the `Reborrow` trait. The
+derived implementation should copy any fields supporting `Copy` and reborrow
+any fields supporting `Reborrow`; if neither is possible for any field then the
+trait cannot be derived.
+
+Anonymous types should have this trait automatically derived as is the case for
+`Copy` today; e.g. `(u32, &i32)` should support `Reborrow`.
+
+## Usage of reborrow
+
+Today, reborrowing happens automatically for references, as in the `f(y)`
+example from the table above. We formalise this:
+
+When a parameter `x` is passed into a function `f` (e.g. `f(x)`),
+
+-   if `x` supports `Copy` [and `x` is later reused], then a copy of `x` will
```

Comment:
I don't think it is necessary to qualify this with `x is later reused`.  A copy is just a move that does not mark the value as uninitialized.